File: src/utils/fs.ts
import { invoke } from '@tauri-apps/api';
import iconv from 'iconv-lite';

export function isDir(path: string): Promise<boolean> {
  return invoke<boolean>('is_dir', { path });
}

export function isFile(path: string): Promise<boolean> {
  return invoke<boolean>('is_file', { path });
}

export function isLink(path: string): Promise<boolean> {
  return invoke<boolean>('is_link', { path });
}

export function exists(path: string): Promise<boolean> {
  return invoke<boolean>('exists', { path });
}

export function mkdir(path: string, isRecursive = true): Promise<boolean> {
  return invoke<boolean>('create_dir', { path, isRecursive });
}

export function copyFile(src: string, dest: string): Promise<boolean> {
  return invoke<boolean>('cp_file', { src, dest });
}

export function copyDir(src: string, dest: string): Promise<boolean> {
  return invoke<boolean>('cp_dir', { src, dest });
}

export function removeFile(path: string): Promise<boolean> {
  return invoke<boolean>('remove_file', { path });
}

export function removeDir(path: string): Promise<boolean> {
  return invoke<boolean>('remove_dir', { path });
}

export function readBuffer(path: string): Promise<Array<number>> {
  return invoke<Array<number>>('read_buffer', { path });
}

export function writeBuffer(
  path: string,
  contents: number[],
): Promise<boolean> {
  return invoke<boolean>('write_buffer', { path, contents });
}

export function appendBuffer(
  path: string,
  contents: Uint8Array,
): Promise<boolean> {
  return invoke<boolean>('append_buffer', { path, contents });
}

export async function readText(path: string, encoding: string = 'utf-8') {
  const bytes = await readBuffer(path);
  const buffer = Buffer.from(bytes);
  return iconv.decode(buffer, encoding);
}

export function writeText(
  path: string,
  text: string,
  encoding: string = 'utf-8',
) {
  const buffer = iconv.encode(text, encoding);
  const bytes = Array.from(buffer);
  return writeBuffer(path, bytes);
}

export function appendText(
  path: string,
  text: string,
  encoding: string = 'utf-8',
) {
  const buffer = iconv.encode(text, encoding);
  return appendBuffer(path, buffer);
}

interface File {
  name: string;
  isDir: boolean;
  isFile: boolean;
  isLink: boolean;
  fullPath: string;
}

type FileStr = File & {
  toString: () => string;
};

export async function childList(path: string): Promise<FileStr[]> {
  const fullPathList = await invoke<string[]>('child_list', { path });

  const result: FileStr[] = [];

  for (const fullPath of fullPathList) {
    const name = fullPath.replaceAll(path, '').slice(1);
    const pathIsDir = await isDir(fullPath);
    const pathIsFile = await isFile(fullPath);
    const pathIsLink = await isLink(fullPath);

    const file = {
      name,
      isDir: pathIsDir,
      isFile: pathIsFile,
      isLink: pathIsLink,
      fullPath,
    };

    result.push({ ...file, toString: () => JSON.stringify(file) });
  }

  return result;
}

export function renameFile(src: string, dest: string): Promise<boolean> {
  return invoke<boolean>('rename_file', { src, dest });
}

File: src-tauri/src/fs.rs
// Learn more about Tauri commands at https://tauri.app/v1/guides/features/command
use std::process::Command;

#[tauri::command]
pub fn is_dir(path: &str) -> bool {
    std::path::Path::new(path).is_dir()
}

#[tauri::command]
pub fn is_file(path: &str) -> bool {
    std::path::Path::new(path).is_file()
}

#[tauri::command]
pub fn is_link(path: &str) -> bool {
    std::path::Path::new(path).is_symlink()
}

#[tauri::command]
pub fn exists(path: &str) -> bool {
    std::path::Path::new(path).exists()
}

#[tauri::command]
pub fn create_dir(path: &str, is_recursive: bool) -> bool {
    if is_recursive {
        std::fs::create_dir_all(path).is_ok()
    } else {
        std::fs::create_dir(path).is_ok()
    }
}

#[tauri::command]
pub fn child_list(path: &str) -> Vec<String> {
    let mut result = Vec::new();
    for entry in std::fs::read_dir(path).unwrap() {
        let entry = entry.unwrap();
        let path = entry.path();
        let path = path.to_str().unwrap();
        result.push(path.to_string());
    }
    result
}

#[tauri::command]
pub fn read_buffer(path: &str) -> Vec<u8> {
    std::fs::read(path).unwrap_or_default()
}

#[tauri::command]
pub fn write_buffer(path: &str, contents: Vec<u8>) -> bool {
    std::fs::write(path, contents).is_ok()
}

#[tauri::command]
pub fn append_buffer(path: &str, contents: Vec<u8>) -> bool {
    use std::fs::OpenOptions;
    use std::io::Write;

    let file: Result<std::fs::File, std::io::Error> =
        OpenOptions::new().append(true).create(true).open(path);

    match file {
        Ok(mut f) => {
            if let Err(err) = f.write_all(&contents) {
                eprintln!("Failed to write to file: {}", err);
                return false;
            }
            true
        }
        Err(err) => {
            eprintln!("Failed to open file: {}", err);
            false
        }
    }
}

#[tauri::command]
pub fn cp_file(src: &str, dest: &str) -> bool {
    std::fs::copy(src, dest).is_ok()
}

#[tauri::command]
pub fn cp_dir(src: &str, dest: &str) -> bool {
    let options = fs_extra::dir::CopyOptions::new();
    return fs_extra::dir::copy(src, dest, &options).is_ok();
}

#[tauri::command]
pub fn remove_dir(path: &str) -> bool {
    std::fs::remove_dir_all(path).is_ok()
}

#[tauri::command]
pub fn remove_file(path: &str) -> bool {
    std::fs::remove_file(path).is_ok()
}

#[tauri::command]
pub fn rename_file(src: &str, dest: &str) -> bool {
    std::fs::rename(src, dest).is_ok()
}

#[tauri::command]
pub fn open_file(path: &str, open_in_folder: bool) {
    let mut command = match std::env::consts::OS {
        "windows" => {
            if open_in_folder {
                Command::new("explorer")
                    .arg("/select,")
                    .arg(path)
                    .spawn()
                    .expect("Failed to execute command")
            } else {
                Command::new("cmd")
                    .args(&["/C", "start", "", path])
                    .spawn()
                    .expect("Failed to execute command")
            }
        }
        "macos" => {
            if open_in_folder {
                Command::new("open")
                    .arg("-R")
                    .arg(path)
                    .spawn()
                    .expect("Failed to execute command")
            } else {
                Command::new("open")
                    .arg(path)
                    .spawn()
                    .expect("Failed to execute command")
            }
        }
        "linux" => {
            if open_in_folder {
                Command::new("xdg-open")
                    .arg("-o")
                    .arg(path)
                    .spawn()
                    .expect("Failed to execute command")
            } else {
                Command::new("xdg-open")
                    .arg(path)
                    .spawn()
                    .expect("Failed to execute command")
            }
        }
        _ => panic!("Unsupported operating system"),
    };

    command.wait().expect("Failed to wait for command");
}
